Sequential Slideshow mode + Image Caching + Next/Previous Image Controls

leogaube opened the pull request:
This is a BIG one...
I am sorry for that, but it's impossible to divide this pull request into smaller ones, because everything seems to be intertwined!
### New Features: ###
**sequential slideshow mode** --> Most of the time it is desirable to show images at random. 
But after going on a vacation, I might want to share my memories with a friend in a sequential order. 
You can select whether you want to "randomize the image order" or not via the web interface.

Preloading the next image used to take about 5-15 seconds depending on internet speed and RaspPi version (for image processing). That's why I introduced 
**image caching** --> All images will be stored inside the "/root/cache" directory. 
Processed images will be stored inside subdirectories "blurred" and "zoomed".
Before using a cached image, I make sure that that image is not corrupted.

**cache garbage collector** --> Of course lots of cached images will never be displayed again if you remove a keyword or even delete some images on Google Drive. For that reason, I created a garbage collector (diskmanager.py) that automatically deletes old files, the "blurred" and "zoomed" subdirectories, or even the entire cache (depending on free/total byte ratio of your SD-card).
The garbage collector runs before the slideshow starts and then after every 100 images.
It is practically impossible to run out of disk space due to cached images (even on small SD-cards).
**NOTE:** If you crop/edit images on GooglePhotos you need to manually delete the cache (web interface) to show these changes.

**slideshow controls** --> If you don't like the current image you can now change it with the click of a button. But maybe the opposite is the case and next image came too early. That's why it is also possible to show any previously displayed image. You can even skip an entire album or go to the previous album (sequential slideshow mode only). The onscreen controls are located over the thumbnail image of the web interface.
By the way, if you want to display a specific image, you can switch to “sequential slideshow mode” and use next/prev Album to easily find the image.

### Changes to the existing code ###
Hopefully I am not overstepping any lines as a contributor, but in order to add these features I had to make some changes by remodeling the communication between slideshow, servicemanager and all services:

**slideshow**
I divided the 'presentation'-method into smaller functions (prepareNextItem, process, delay, show, repeat) so that it becomes more human readable. There also used to be a lot of redundant calls.
'imageCurrent' used to be the preloaded image, it now really is the currently displayed image.

**servicemanager**
Now the serviceManager is solely responsible which service gets used when (not the slideshow). 
The serviceManager also handles service-related errors (not the service itself). 
It remembers which services have already displayed all its images (random image order only). That means 'memoryForget' is only called if all services are out of images. 
**NOTE:** Currently every service is still picked with the same probability. So, if you have one service with 1000 images and another with only 10, the second service will run out of images really soon.
Fixing this issue is already on my ToDo-list (by counting the images of each service), but it seems to be better than displaying these 10 images repeatedly until the first service runs out of images as well.

**svc_xxx**
I added more service states; thus a service is only considered as "ready" if e.g. keywords are supplied.
An image will only be "remembered" if it has been displayed (not only preloaded).
A "simple url" service is treated as a single album that way you can have multiple services each grouping related content.

### Conclusion ###
It might seem like a lot has changed, but most of these changes are only little performance improvements and don't affect the functionality of the photoframe. Just give it a try.
I did a lot of testing to make sure this update doesn't break anything, but you can never be sure...
If you have any questions about new features or my changes, please let me know! 
Once again thank you very much for making this project open-source :)

Cheers,
Leo

mrworf commented:
Very interesting stuff. I'm currently out of town so I could only do a quick look on my phone (less than ideal) and it looks pretty good. There are some things we should discuss but I look forward to being able to incorporate this into mainline.

mrworf commented on modules/helper.py:
```diff
@@ -58,23 +58,73 @@ def getExtension(mime):
+		return True
+		
+
+	# can also be used to check if image file is corrupted
```
This function doesn't make me super happy. It has an odd behavior where it will actually delete files if asked to. It's an odd extra job for a function which should only return the size. Worse, the caller won't know if that happened since it is magic 😄 

It's used predominantly by the various content providers, but for them, it would make more sense to do a `os.unlink` on the file in-question by default to make sure it's not there. Or simply allow the request function to overwrite any existing content.

leogaube replied:
You are right, as always ;)
The function doesn't delete files anymore! I moved that part to _useCachedImage_ of the baseService

mrworf commented on services/svc_googlephotos.py:
```diff
@@ -139,50 +141,50 @@ def addKeywords(self, keywords):
-
-  def fetchImage(self, destinationFile, supportedMimeTypes, displaySize):
-    # First, pick which keyword to use
+  def prepareNextItem(self, destinationDir, supportedMimeTypes, displaySize, randomize):
```
It feels like this function has gotten a bit more complicated (random vs sequential plus caching). These are great ideas, but by forcing all services to implement them, the risk for bugs have gone up.

If the `base.py` implemented a method for `getNextKeyword()` which did this random vs sequental stuff, it would make it less likely that someone (ie, me 😜 ) would screw up on writing a new service or changing one that exists.

Likewise, by putting the caching logic into the diskmanager class (which essentially is the cache manager?) and simply having a `getCachedCopy()` function (which returns `None` if none exists), there is no need to call the `getImageSize` function and the flow would make a bit more sense.

leogaube replied:
I did my best to create a general flow for selecting an album/image. 
Now if you want to add a new service that uses keywords you only need to override `getImagesFor(keyword)` and `addUrlParams`.
https://github.com/mrworf/photoframe/blob/c71f40d6a673ea9622298bd8ae632fe8a52633f6/services/base.py#L370-L392

Everything else should be handled by `prepareNextItem`-->`selectImageFromAlbum`-->`selectImage` of the baseService!

It is still possible to return service-specific error messages, for example:
https://github.com/mrworf/photoframe/blob/c71f40d6a673ea9622298bd8ae632fe8a52633f6/services/svc_googlephotos.py#L242-L250

mrworf commented on services/svc_googlephotos.py:
```diff
@@ -192,16 +194,26 @@ def isGooglePhotosEnabled(self):
     return not (data['status'] == 403 and 'Enable it by visiting' in data['content'])
 
-  def getUrlFromImages(self, types, displaySize, images):
+  def getUrlFromImages(self, types, displaySize, images, randomize):
```
Yes, obviously, given the previous comment, this will obvious need to be handled specially unless you can think of a way to abstract this behavior into the `base.py`

leogaube replied:
I changed `getUrlFromImages` to a more abstract `selectImage` of `base.py`

mrworf commented on services/svc_simpleurl.py:
```diff
@@ -23,25 +27,65 @@ class SimpleUrl(BaseService):
 
-
-  def prepareNextItem(self, destinationFile, supportedMimeTypes, displaySize):
+  def prepareNextItem(self, destinationDir, supportedMimeTypes, displaySize, randomize):
```
This one would certainly benefit from the abstraction mentioned above since it grew quite a bit to support the new features.

leogaube replied:
yes it does. 
I kept it to distinguish between services that `needKeywords()` (-->`selectImageFromAlbum`) and services that don't (no such service implemented yet).

mrworf commented on services/base.py:
```diff
@@ -414,4 +478,19 @@ def isCorrectOrientation(self, metadata, displaySize):
 
     return image_orientation == display_orientation
 
+  def nextAlbum(self):
```
Comment to explain for newcomers what this does

leogaube replied:
done

mrworf commented on services/base.py:
```diff
@@ -414,4 +478,19 @@ def isCorrectOrientation(self, metadata, displaySize):
+      return False
+    return True
+
+  def prevAlbum(self):
```
ditto

leogaube replied:
done

mrworf commented on static/js/main.js:
```diff
@@ -1,69 +1,96 @@
 
 function rebootWatch() {
   $(document.body).html('<h1>Rebooting</h1>')
+  setTimeout(document.location.reload.bind(document.location), 60000);
```
Not foolproof, especially if running a slow overworked RPiZW with shitty SD but certainly nicer than nothing 😸

mrworf commented on static/js/main.js:
```diff
@@ -221,6 +274,13 @@ $("input[class='keyword-delete']").click(function(){
 
+$('.keyword').keypress(function (e) {
+  if (e.which == 13) {
+    $(this).next(".keyword-add").trigger("click");
```
nice

leogaube replied:
:)

leogaube commented:
> But I like the ideas and changes and the slideshow controls are slick, so I hope this review doesn't discourage you.

No, this review doesn't discourage me at all. On the contratry, I am really glad for all the critique and I agree with almost every issue (big and small)! 

> * Try to avoid duplicate functionality in each service, put as much general functionality in the base.py to help newcomers.

That part was always on the back of my mind too, I just need to come up with a good strategy on how to approch the issue.

> * While PicasaWeb isn't going to be supported anymore, people may still be using it (not tested it in a while) and your changes will most likely break all frames using that code, so it needs to be revised (if it's finally dead, let me know and we'll figure something out. Haven't tried it in a while so who knows).

The problem is that I never used Picassa, thus I don't have a way to test new code!
However, their website clearly states that PicassaWeb is not only depricated but also "*shut down, including all the end-points.*" (https://developers.google.com/picasa-web/docs/3.0/deprecation)
So maybe it is time for you to remove it completely?



I just started a new job this week, but hopefully I'll get around to all of your proposed changes on the weekend :)

leogaube commented on frame.py:
```diff
@@ -577,6 +599,7 @@ def services_operations(action):
 timekeeper = timekeeper(display.enable, slideshow.start)
 slideshow.setQueryPower(timekeeper.getDisplayOn)
 slideshow.setServiceManager(services)
+slideshowEvents = ["nextImage", "prevImage", "nextAlbum", "prevAlbum"]
```
yes, that makes more sense

leogaube commented on frame.py:
```diff
@@ -521,6 +533,16 @@ def services_operations(action):
+@auth.login_required
+def control_slideshow(cmd):
+  if cmd in slideshowEvents:
+    slideshow.createEvent(cmd)
```
I changed it

leogaube commented on modules/diskmanager.py:
```diff
@@ -0,0 +1,148 @@
+MB = KB * 10**3
+GB = MB * 10**3
+
+def formatBytes(size):
```
I made _*formatBytes*_ part of the class, but doing the same for all the constants would imply that you would have to write Cachemanager.MB, Cachemanager.HOUR, etc. (which is annoying and does not make much sense to me)

leogaube replied:
if you want to use these constants from a different file you could always use 
`from modules.cachemanager import MB, GB, MONTH ...`

or I could place them inside a completly different file: _constants.py_

leogaube commented on modules/diskmanager.py:
```diff
@@ -0,0 +1,148 @@
+        if stat.st_mtime < now - minAge:
+          logging.debug("deleting old file '%s'"%filename)
+          freedUpSpace += stat.st_size
+          os.remove(filename)
```
I fixed it

leogaube commented on modules/helper.py:
```diff
@@ -58,23 +58,73 @@ def getExtension(mime):
+
+		cmd = ["/usr/bin/file", "--mime", filename]
+		try:
+			process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
yes, it would. I fixed it

leogaube commented on modules/servicemanager.py:
```diff
@@ -295,12 +330,86 @@ def getServices(self, readyOnly=False):
+      svc = self.forceService
+    elif randomize:
+      availableServices = [s for s in availableServices if self._SERVICES[s['id']]['service'] not in self._OUT_OF_IMAGES]
+      logging.debug("# of avaialable services %d"%len(availableServices))
```
done

leogaube commented on modules/slideshow.py:
```diff
@@ -69,95 +87,199 @@ def start(self, blank=False):
+            msg += "add one or more keywords (album names)"
+          elif state == 'NO_IMAGES_DETECTED':
+            msg += "no images could be found!"
+            if svcName == 'USB-Photos':
```
You're right. I moved that part to an optional explainState function.

leogaube commented on modules/slideshow.py:
```diff
@@ -69,95 +87,199 @@ def start(self, blank=False):
+            msg += "no images could be found!"
+            if svcName == 'USB-Photos':
+              msg += "\nPlace images and/or albums inside a '/photoframe'-directory on your storage device"
+          elif state == 'CONNECT_STORAGE_DEVICE':
```
ditto

leogaube commented on modules/slideshow.py:
```diff
@@ -69,95 +87,199 @@ def start(self, blank=False):
+    if self.colormatch.hasSensor():
+      # For Now: Always process original image (no caching of colormatch-adjusted images)
+      # 'colormatched_tmp.jpg' will be deleted after the image is displayed
+      filenameTemp = os.path.join(self.settings.getUser('cachefolder'), "colormatched_tmp.jpg")
```
OK

leogaube commented on services/base.py:
```diff
@@ -240,10 +248,18 @@ def getConfigurationFields(self):
     return {'error':None, 'keywords': keywords}
 
-  def addKeywords(self, keywords):
+  def addKeywords(self, keywords): #shouldn't this be 'addKeyword' (singular)? seems a bit confusing to me, because 'keywords' is a string and not a list!
```
I left it as it is for now. Maybe you can change it whenever you decide to remove PicassaWeb

leogaube commented on services/base.py:
```diff
@@ -380,29 +396,77 @@ def _fetchMemory(self, key):
     return h in self._MEMORY
 
-  def memoryForget(self, keywords=None):
+  def memoryForgetLast(self, keywords=None):
```
done

leogaube replied:
I also added a comment inside _\_\_init\_\__ explaining the difference between MEMORY and HISTORY

leogaube commented on services/base.py:
```diff
@@ -380,29 +396,77 @@ def _fetchMemory(self, key):
+      self._HISTORY = []
+
+  def getPreferredImageSize(self, imageSize, displaySize):
+    # Calculate the size we need to avoid black borders
```
It's a little hard to explain, but I tried my best with the comment:
https://github.com/mrworf/photoframe/blob/c71f40d6a673ea9622298bd8ae632fe8a52633f6/services/base.py#L520-L525

It basically returns the smallest possible size, so that no upscaling needs to be done during _processing_ (regardless of scaling option).

leogaube commented:
Hey Happy Easter :)
I tried to fix everything, but please let me know if there is still anything that you dislike!

mrworf commented on services/base.py:
```diff
@@ -39,6 +40,9 @@ class BaseService:
   STATE_DO_OAUTH = 2
+  STATE_NEED_KEYWORDS = 3
+  STATE_NO_IMAGES = 4
+  STATE_NOT_CONNECTED = 5
```
Not needed anymore, right?

leogaube replied:
yes

leogaube commented:
Hi, 
I am really sorry that it took me so long to fix these minor objections. I was out of town and than I just couldn't find any time last week...

Hopefully, I didn't misunderstand you regarding the new states NEED_KEYWORDS, NO_IMAGES, NOT_CONNECTED:
I totally understand that you want `STATE_NOT_CONNECTED` only as a substate of the usb-service, but is my assumption correct that you are fine with the first two being part of the base class?

In order for `STATE_NO_IMAGES` to work for all services, I added a variable (`base._STATE['_NUM_IMAGES']`) that keeps track of how many images are provided by each album/service. That way, the process for selecting a random service/album can also be weighted by the #images provided by each service/album. Therefore, every single image has the same chance of being displayed. 
This should improve the "randomness" in some use-cases e.g. a photoframe with two services, one with few images and the other with a lot of images. 
As for now, unsupported images (mimetype, orientation) and already displayed images are NOT excluded from NUM_IMAGES due to simplicity, but it should still serve as a good estimate (and it definitely works better than before)
Additionally, a warning can be displayed on the web interface for keywords that don't provide any images (e.g. due to spelling, broken url, ...)

I also noticed that an unstable internet connection or a single bad url (e.g. from **simple_url**) causes the photoframe to freeze due to an unhandled exception from `base.requestUrl()`. Therefore, I added a `keyword_validation()` for **simple_url** that does not allow malformed urls (e.g. missing "http") and I also catch `requestUrl()`-errors, trying to differentiate between an unavailable service or a lack of internet connection!

leogaube commented on modules/servicemanager.py:
```diff
@@ -245,12 +261,28 @@ def getServiceState(self, id):
+      return 'NEED_KEYWORDS'
+    elif state == svc.STATE_NO_IMAGES:
+      return 'NO_IMAGES_DETECTED'
+    elif state == svc.STATE_NOT_CONNECTED:
```
yes

leogaube commented on modules/slideshow.py:
```diff
@@ -165,43 +170,44 @@ def handleErrors(self, result):
         msg = 'Photoframe isn\'t ready yet\n\nPlease direct your webbrowser to\n\nhttp://%s:7777/\n\nand add one or more photo providers' % self.settings.get('local-ip')
       else:
         msg = 'Please direct your webbrowser to\n\nhttp://%s:7777/\n\nto complete the setup process:\n\n' % self.settings.get('local-ip')
```
Actually, that was intensional, because it looked better in my opinion, but I changed it.

leogaube commented on modules/slideshow.py:
```diff
@@ -165,43 +170,44 @@ def handleErrors(self, result):
+              msg += additionalInfo
+              additionalInfo = None
+
+          if additionalInfo is not None:
```
I had something in mind for future work, but it is definitely really confusing, so I removed it.

mrworf commented on services/base.py:
```diff
@@ -440,7 +477,14 @@ def selectImageFromAlbum(self, destinationDir, supportedMimeTypes, displaySize,
 
-      result = self.requestUrl(url, destination=filename)
+      try:
+        result = self.requestUrl(url, destination=filename)
```
Just a side-note, we might want to do this inside the requestUrl function instead and simply return None if the request cannot be handled. This has the benefit that anyone using the requestUrl() call will automatically get the exception handling. But this can be done as a TODO since I'd rather have this massive PR done 😄

leogaube replied:
You're right. I didn't want to do it inside the requestUrl function, because I was not sure if some future service might want to differentiate between different errors (max retries exceeded, request timeout, ...), especially with POST functionallity already built-in! I just didn't want to mess that up, but feel free to change it :)

As for the _"massive PR"_: I'll try to keep my next PRs small and simple ;)

mrworf commented:
Oh, also, you may want to merge the changes on master onto your branch. I had to do some fixes for failing frames a week or so ago. May actually fix the `requestUrl()` exception handling code already.
